File: Reference/src/Brontide.Reference.Experimental.Binding/Portable/PortableBindingPlan.cs
using System.Collections.Immutable;
using System.Globalization;

namespace Brontide.Reference.Experimental.Binding.Portable;

/// <summary>The declared payload representations.</summary>
public static class PortableRepresentations
{
    /// <summary>The portable wire for the negotiated process realization.</summary>
    public const string PortableCborCore = "portable-cbor-core-0.1";

    /// <summary>
    /// The retained Cooling and Catalog representation. It is diagnostic and legacy only and is not
    /// the portable wire contract, so a portable binding refuses it rather than negotiating it.
    /// </summary>
    public const string InlineTaggedJson = "inline-tagged-json";

    public const string LengthDelimited = "length-delimited";
    public const string DirectCall = "direct-call";
}

/// <summary>
/// One immutable, inspectable Binding Plan for one binding scope.
/// </summary>
/// <remarks>
/// Every fact is fixed when negotiation succeeds and never changes. A change requires a new binding
/// scope with a new plan identifier: there is no renegotiation in place. Each fact is also
/// retrievable by name through <see cref="Fact"/> without invoking the provider, so a compiled
/// realization answers the same inspection questions as an explicit-data one.
/// </remarks>
public sealed record PortableBindingPlan
{
    private readonly ImmutableSortedDictionary<string, string> _facts;

    internal PortableBindingPlan(
        PortablePlanId planId,
        PortableContractDocument contract,
        ImmutableArray<PortableOperationReference> operations,
        ImmutableArray<PortableShapeReference> shapes,
        ImmutableArray<PortableFragmentReference> fragments,
        ImmutableArray<PortableDependencyReference> satisfiedRequirements,
        ImmutableArray<PortableCompactAssignment> compactIdentifierAssignments,
        string hostEndpoint,
        string providerEndpoint,
        string selectionReason,
        PortableRealization realization)
    {
        PlanId = planId;
        Contract = contract;
        Operations = operations;
        Shapes = shapes;
        Fragments = fragments;
        SatisfiedRequirements = satisfiedRequirements;
        CompactIdentifierAssignments = compactIdentifierAssignments;
        HostEndpoint = hostEndpoint;
        ProviderEndpoint = providerEndpoint;
        SelectionReason = selectionReason;
        Realization = realization;
        Catalog = PortableShapeCatalog.FromContract(contract);
        _facts = BuildFacts();
    }

    public PortablePlanId PlanId { get; }

    public PortableContractDocument Contract { get; }

    public PortableShapeCatalog Catalog { get; }

    public int ContractVersion => Contract.ContractVersion;

    public PortableComponentReference Component => Contract.Component;

    public PortableProviderReference Provider => Contract.Provider;

    public ImmutableArray<PortableOperationReference> Operations { get; }

    public ImmutableArray<PortableShapeReference> Shapes { get; }

    public ImmutableArray<PortableFragmentReference> Fragments { get; }

    public ImmutableArray<PortableDependencyReference> SatisfiedRequirements { get; }

    public ImmutableArray<PortableCompactAssignment> CompactIdentifierAssignments { get; }

    public string HostEndpoint { get; }

    public string ProviderEndpoint { get; }

    public PortableProviderReference SelectedProvider => Contract.Provider;

    public string SelectionReason { get; }

    public PortableAuthorityMode AuthorityPresentationMode => Contract.Authority.PresentationMode;

    public bool TrustBoundaryCrossed => Contract.Authority.TrustBoundaryCrossed;

    public bool NoCapabilityTransfer => Contract.Authority.NoCapabilityTransfer;

    public PortableAuthorityDecisionPoint AuthorityDecisionPoint =>
        AuthorityPresentationMode == PortableAuthorityMode.CrossTrustNoCapabilityTransfer
            ? PortableAuthorityDecisionPoint.ProviderDomain
            : PortableAuthorityDecisionPoint.TargetBoundary;

    /// <summary>
    /// Cross-trust revocability is carried-tier by default; a resolver or indirection point is a
    /// declared capability, never an assumed one.
    /// </summary>
    public string ChainConjunctionTier => "carried";

    public string Representation => Contract.Representation.Representation;

    public string Framing => Realization == PortableRealization.FixedDirectCall
        ? PortableRepresentations.DirectCall
        : Contract.Representation.Framing;

    public ImmutableArray<string> ResourceFlavors => Contract.Representation.ResourceFlavors;

    public ImmutableArray<string> AcceptedResourceHandles => Contract.Representation.AcceptedResourceHandles;

    public string ResourceOwnership => ResourceFlavors.IsEmpty
        ? "none"
        : ResourceFlavors.Contains(PortableResourceFlavors.CopiedImmutableBlob, StringComparer.Ordinal)
            ? "copied"
            : "provider-retained";

    public bool ImplicitCopyPermitted => false;

    public string? IntegrityAlgorithm =>
        ResourceFlavors.Contains(PortableResourceFlavors.CopiedImmutableBlob, StringComparer.Ordinal) ? "SHA-256" : null;

    public int MaxConcurrentRequests => Contract.Limits.MaxConcurrentRequests;

    public string OrderingGuarantee => "none";

    public PortableRealization Realization { get; }

    public ImmutableArray<string> CrossedBoundaries
    {
        get
        {
            var boundaries = ImmutableArray.CreateBuilder<string>();
            if (Realization == PortableRealization.NegotiatedProcess)
            {
                boundaries.Add("process");
            }

            if (TrustBoundaryCrossed)
            {
                boundaries.Add("trust");
            }

            return boundaries.Count == 0 ? ["none"] : boundaries.ToImmutable();
        }
    }

    public PortableLimits Limits => Contract.Limits;

    public bool ReplayProtectionDeclared => Contract.Lifecycle.ReplayProtectionDeclared;

    public string ReplayWindow => Contract.Lifecycle.ReplayWindow;

    public ImmutableSortedDictionary<string, bool> LifecycleFeatures => Contract.Lifecycle.Features;

    public ImmutableArray<string> TerminalStates { get; } = ["terminated", "failed"];

    public ImmutableArray<string> FailureDomainsObservable { get; } =
    [
        "local-endpoint", "transport", "remote-endpoint", "remote-provider", "unknown"
    ];

    public bool ExceptionTransportPermitted => false;

    public IEnumerable<string> FactNames => _facts.Keys;

    /// <summary>Retrieves one plan fact by name without invoking the provider.</summary>
    public string Fact(string name)
    {
        ArgumentNullException.ThrowIfNull(name);
        return _facts.TryGetValue(name, out var value)
            ? value
            : throw new ArgumentOutOfRangeException(nameof(name), $"The Binding Plan declares no fact named '{name}'.");
    }

    public IReadOnlyDictionary<string, string> Facts => _facts;

    public PortableOperationDeclaration Operation(PortableOperationReference reference) =>
        Contract.Operations.FirstOrDefault(operation => operation.Reference == reference) ??
        throw new PortableFaultException(
            PortableProtocolCategory.UnsupportedOperation,
            "unknown-operation",
            $"Operation {reference} is outside the established contract.");

    private ImmutableSortedDictionary<string, string> BuildFacts()
    {
        var builder = ImmutableSortedDictionary.CreateBuilder<string, string>(StringComparer.Ordinal);
        builder["planId"] = PlanId.Value;
        builder["contractVersion"] = ContractVersion.ToString(CultureInfo.InvariantCulture);
        builder["component"] = Component.ToString();
        builder["provider"] = Provider.ToString();
        builder["operations"] = Join(Operations.Select(operation => operation.ToString()));
        builder["shapes"] = Join(Shapes.Select(shape => shape.ToString()));
        builder["fragments"] = Join(Fragments.Select(fragment => fragment.ToString()));
        builder["satisfiedRequirements"] = Join(SatisfiedRequirements.Select(requirement => requirement.ToString()));
        builder["compactIdentifierAssignments"] = Join(CompactIdentifierAssignments
            .Select(assignment =>
                $"{IdentitySpaceToken(assignment.Space)}:{assignment.Reference}={assignment.CompactId}"));
        builder["hostEndpoint"] = HostEndpoint;
        builder["providerEndpoint"] = ProviderEndpoint;
        builder["selectedProvider"] = SelectedProvider.ToString();
        builder["selectionReason"] = SelectionReason;
        builder["authorityPresentationMode"] = AuthorityPresentationMode.Token();
        builder["trustBoundaryCrossed"] = Render(TrustBoundaryCrossed);
        builder["noCapabilityTransfer"] = Render(NoCapabilityTransfer);
        builder["constraintPolicy"] = Contract.Authority.ConstraintPolicy;
        builder["authorityDecisionPoint"] = AuthorityDecisionPoint.Token();
        builder["chainConjunctionTier"] = ChainConjunctionTier;
        builder["representation"] = Representation;
        builder["framing"] = Framing;
        builder["resourceFlavors"] = Join(ResourceFlavors);
        builder["acceptedResourceHandles"] = Join(AcceptedResourceHandles);
        builder["resourceOwnership"] = ResourceOwnership;
        builder["implicitCopyPermitted"] = Render(ImplicitCopyPermitted);
        builder["integrityAlgorithm"] = IntegrityAlgorithm ?? "absent";
        builder["maxConcurrentRequests"] = MaxConcurrentRequests.ToString(CultureInfo.InvariantCulture);
        builder["orderingGuarantee"] = OrderingGuarantee;
        builder["realization"] = Realization.Token();
        builder["crossedBoundaries"] = Join(CrossedBoundaries);
        builder["limits"] = Join(
        [
            $"maxFrameBytes={Limits.MaxFrameBytes}",
            $"maxNestingDepth={Limits.MaxNestingDepth}",
            $"maxRecordFields={Limits.MaxRecordFields}",
            $"maxFragmentsPerRecord={Limits.MaxFragmentsPerRecord}",
            $"maxSequenceItems={Limits.MaxSequenceItems}",
            $"maxTextBytes={Limits.MaxTextBytes}",
            $"maxByteStringBytes={Limits.MaxByteStringBytes}",
            $"maxResourceBytes={Limits.MaxResourceBytes}",
            $"ioTimeoutMilliseconds={Limits.IoTimeoutMilliseconds}",
            $"maxConcurrentRequests={Limits.MaxConcurrentRequests}"
        ]);
        builder["replayProtectionDeclared"] = Render(ReplayProtectionDeclared);
        builder["replayWindow"] = ReplayWindow;
        builder["lifecycleFeatures"] = Join(LifecycleFeatures.Select(feature => $"{feature.Key}={Render(feature.Value)}"));
        builder["terminalStates"] = Join(TerminalStates);
        builder["failureDomainsObservable"] = Join(FailureDomainsObservable);
        builder["exceptionTransportPermitted"] = Render(ExceptionTransportPermitted);
        return builder.ToImmutable();
    }

    private static string IdentitySpaceToken(PortableIdentitySpace space) => space switch
    {
        PortableIdentitySpace.Operation => "operation",
        PortableIdentitySpace.Shape => "shape",
        PortableIdentitySpace.Fragment => "fragment",
        _ => throw new ArgumentOutOfRangeException(nameof(space)),
    };

    private static string Join(IEnumerable<string> values) => string.Join(", ", values);

    private static string Render(bool value) => value ? "true" : "false";
}

File: Minimal/src/Brontide.Minimal.Binding/Portable/PortableBindingPlan.fs
namespace Brontide.Minimal.Binding.Portable

/// One immutable, inspectable Binding Plan for one binding scope.
///
/// Every fact is fixed when negotiation succeeds and never changes. A change requires a new binding
/// scope with a new plan identifier: there is no renegotiation in place. Each fact is also
/// retrievable by name without invoking the provider, so a compiled realization answers the same
/// inspection questions as an explicit-data one.
[<StructuralEquality; NoComparison>]
type BindingPlan =
    private
        { Identifier: PlanId
          Contract: ContractDocument
          AnsweringProvider: PortableProviderRef
          Catalog: ShapeCatalog
          Operations: PortableOperationRef list
          Shapes: PortableShapeRef list
          Fragments: PortableFragmentRef list
          SatisfiedRequirements: PortableDependencyRef list
          CompactAssignments: CompactAssignment list
          HostEndpoint: string
          ProviderEndpoint: string
          SelectionReason: string
          Realization: Realization
          Facts: Map<string, string> }

[<RequireQualifiedAccess>]
module BindingPlan =

    let private render (flag: bool) = if flag then "true" else "false"

    let private join (values: string seq) = System.String.Join(", ", values)

    let planId plan = plan.Identifier
    let contract plan = plan.Contract
    let catalog plan = plan.Catalog
    let contractVersion plan = plan.Contract.ContractVersion
    let component' plan = plan.Contract.Component
    /// The provider that answered, read from the offered document.
    ///
    /// Negotiation refuses a provider mismatch, so this equals the required declaration whenever a
    /// plan exists. Reading it from the offered side keeps the fact honest about who answered rather
    /// than who was asked, which is what a field of this name has to mean. Decision 11.
    let answeringProvider plan = plan.AnsweringProvider

    let provider plan = plan.AnsweringProvider
    let operations plan = plan.Operations
    let shapes plan = plan.Shapes
    let fragments plan = plan.Fragments
    let satisfiedRequirements plan = plan.SatisfiedRequirements
    let compactAssignments plan = plan.CompactAssignments
    let hostEndpoint plan = plan.HostEndpoint
    let providerEndpoint plan = plan.ProviderEndpoint
    let selectedProvider plan = plan.AnsweringProvider
    let selectionReason plan = plan.SelectionReason
    let realization plan = plan.Realization
    let limits plan = plan.Contract.Limits
    let authority plan = plan.Contract.Authority
    let trustBoundaryCrossed plan = plan.Contract.Authority.TrustBoundaryCrossed
    let representation plan = plan.Contract.Representation.Representation
    let resourceFlavors plan = plan.Contract.Representation.ResourceFlavors
    let acceptedResourceHandles plan = plan.Contract.Representation.AcceptedResourceHandles
    let replayProtectionDeclared plan = plan.Contract.Lifecycle.ReplayProtectionDeclared
    let replayWindow plan = plan.Contract.Lifecycle.ReplayWindow
    let lifecycleFeatures plan = plan.Contract.Lifecycle.Features

    /// A direct call has no wire, so its framing is the realization itself rather than the
    /// length-delimited framing the contract declares for a process seam.
    let framing plan =
        match plan.Realization with
        | Realization.FixedDirectCall -> PortableRepresentations.DirectCall
        | Realization.NegotiatedProcess -> plan.Contract.Representation.Framing

    let authorityDecisionPoint plan =
        match plan.Contract.Authority.PresentationMode with
        | AuthorityMode.CrossTrustNoCapabilityTransfer -> AuthorityDecisionPoint.ProviderDomain
        | AuthorityMode.LocalDomainEvaluated -> AuthorityDecisionPoint.TargetBoundary

    /// Cross-trust revocability is carried-tier by default; a resolver or indirection point is a
    /// declared capability, never an assumed one.
    let chainConjunctionTier _ = "carried"

    let resourceOwnership plan =
        let flavors = plan.Contract.Representation.ResourceFlavors

        if List.isEmpty flavors then "none"
        elif List.contains ResourceFlavor.CopiedImmutableBlobToken flavors then "copied"
        else "provider-retained"

    let implicitCopyPermitted _ = false

    let integrityAlgorithm plan =
        if List.contains ResourceFlavor.CopiedImmutableBlobToken plan.Contract.Representation.ResourceFlavors then
            Some "SHA-256"
        else
            None

    let crossedBoundaries plan =
        let boundaries =
            [ if plan.Realization = Realization.NegotiatedProcess then
                  "process"
              if plan.Contract.Authority.TrustBoundaryCrossed then
                  "trust" ]

        if List.isEmpty boundaries then [ "none" ] else boundaries

    let terminalStates = [ "terminated"; "failed" ]

    let failureDomainsObservable =
        FailureDomain.all |> List.map FailureDomain.token

    let exceptionTransportPermitted _ = false

    let tryFact name plan = Map.tryFind name plan.Facts

    let factNames plan =
        plan.Facts |> Map.toList |> List.map fst

    let operation reference plan : PortableResult<OperationDeclaration> =
        match ContractDocument.tryOperation reference plan.Contract with
        | Some declaration -> Ok declaration
        | None ->
            refuse
                ProtocolCategory.UnsupportedOperation
                "unknown-operation"
                $"Operation {PortableOperationRef.text reference} is outside the established contract."

    let private buildFacts plan =
        let limits = plan.Contract.Limits

        Map.ofList
            [ "planId", PlanId.value plan.Identifier
              "contractVersion", string plan.Contract.ContractVersion
              "component", PortableComponentRef.text plan.Contract.Component
              "provider", PortableProviderRef.text plan.AnsweringProvider
              "operations", join (plan.Operations |> List.map PortableOperationRef.text)
              "shapes", join (plan.Shapes |> List.map PortableShapeRef.text)
              "fragments", join (plan.Fragments |> List.map PortableFragmentRef.text)
              "satisfiedRequirements", join (plan.SatisfiedRequirements |> List.map PortableDependencyRef.text)
              "compactIdentifierAssignments",
              join (
                  plan.CompactAssignments
                  |> List.map (fun assignment ->
                      $"{IdentitySpace.token assignment.Space}:{assignment.Reference}={CompactId.value assignment.Compact}")
              )
              "hostEndpoint", plan.HostEndpoint
              "providerEndpoint", plan.ProviderEndpoint
              "selectedProvider", PortableProviderRef.text (selectedProvider plan)
              "selectionReason", plan.SelectionReason
              "authorityPresentationMode", AuthorityMode.token plan.Contract.Authority.PresentationMode
              "trustBoundaryCrossed", render plan.Contract.Authority.TrustBoundaryCrossed
              "noCapabilityTransfer", render plan.Contract.Authority.NoCapabilityTransfer
              "constraintPolicy", plan.Contract.Authority.ConstraintPolicy
              "authorityDecisionPoint", AuthorityDecisionPoint.token (authorityDecisionPoint plan)
              "chainConjunctionTier", chainConjunctionTier plan
              "representation", plan.Contract.Representation.Representation
              "framing", framing plan
              "resourceFlavors", join plan.Contract.Representation.ResourceFlavors
              "acceptedResourceHandles", join plan.Contract.Representation.AcceptedResourceHandles
              "resourceOwnership", resourceOwnership plan
              "implicitCopyPermitted", render (implicitCopyPermitted plan)
              "integrityAlgorithm", (integrityAlgorithm plan |> Option.defaultValue "absent")
              "maxConcurrentRequests", string limits.MaxConcurrentRequests
              "orderingGuarantee", "none"
              "realization", Realization.token plan.Realization
              "crossedBoundaries", join (crossedBoundaries plan)
              "limits",
              join
                  [ $"maxFrameBytes={limits.MaxFrameBytes}"
                    $"maxNestingDepth={limits.MaxNestingDepth}"
                    $"maxRecordFields={limits.MaxRecordFields}"
                    $"maxFragmentsPerRecord={limits.MaxFragmentsPerRecord}"
                    $"maxSequenceItems={limits.MaxSequenceItems}"
                    $"maxTextBytes={limits.MaxTextBytes}"
                    $"maxByteStringBytes={limits.MaxByteStringBytes}"
                    $"maxResourceBytes={limits.MaxResourceBytes}"
                    $"ioTimeoutMilliseconds={limits.IoTimeoutMilliseconds}"
                    $"maxConcurrentRequests={limits.MaxConcurrentRequests}" ]
              "replayProtectionDeclared", render plan.Contract.Lifecycle.ReplayProtectionDeclared
              "replayWindow", plan.Contract.Lifecycle.ReplayWindow
              "lifecycleFeatures",
              join (
                  plan.Contract.Lifecycle.Features
                  |> Map.toList
                  |> List.map (fun (name, declared) -> $"{name}={render declared}")
              )
              "terminalStates", join terminalStates
              "failureDomainsObservable", join failureDomainsObservable
              "exceptionTransportPermitted", render (exceptionTransportPermitted plan) ]

    /// Freezes one plan. Only negotiation calls this, because a plan that was not negotiated has no
    /// facts to be frozen from.
    let internal freeze
        (identifier: PlanId)
        (document: ContractDocument)
        (answeringProvider: PortableProviderRef)
        (catalog: ShapeCatalog)
        (operations: PortableOperationRef list)
        (shapes: PortableShapeRef list)
        (fragments: PortableFragmentRef list)
        (satisfied: PortableDependencyRef list)
        (assignments: CompactAssignment list)
        (hostEndpoint: string)
        (providerEndpoint: string)
        (selectionReason: string)
        (realization: Realization)
        =
        let plan =
            { Identifier = identifier
              Contract = document
              AnsweringProvider = answeringProvider
              Catalog = catalog
              Operations = operations
              Shapes = shapes
              Fragments = fragments
              SatisfiedRequirements = satisfied
              CompactAssignments = assignments
              HostEndpoint = hostEndpoint
              ProviderEndpoint = providerEndpoint
              SelectionReason = selectionReason
              Realization = realization
              Facts = Map.empty }

        { plan with Facts = buildFacts plan }
